cdc/puller/unified_sorter.go

```diff
@@ -0,0 +1,832 @@
+			freedCount := 0
+			for i := range ret.cache {
+				ptr := &ret.cache[i]
+				innerPtr := atomic.SwapPointer(ptr, nil)
+				if innerPtr == nil {
+					continue
+				}
+				backEnd := (*fileSorterBackEnd)(innerPtr)
+				err := backEnd.free()
+				if err != nil {
+					log.Warn("Cannot remove temporary file for sorting", zap.String("file", backEnd.name), zap.Error(err))
+				} else {
+					log.Info("Temporary file removed", zap.String("file", backEnd.name))
+					freedCount += 1
+				}
+				if freedCount >= 16 {
+					freedCount = 0
+					break
+				}
+			}
+		}
+	}()
+
+	return ret
+}
+
+func (p *backEndPool) alloc() (sorterBackEnd, error) {
+	if atomic.LoadInt32(&p.memPressure) < memoryPressureThres {
+		ret := new(memorySorterBackEnd)
+		atomic.AddInt64(&p.memoryUseEstimate, heapSizeLimit)
+		return ret, nil
+	}
+
+	for i := range p.cache {
+		ptr := &p.cache[i]
+		ret := atomic.SwapPointer(ptr, nil)
+		if ret != nil {
+			return (*fileSorterBackEnd)(ret), nil
+		}
+	}
+
+	fname := fmt.Sprintf("%s/sort-%d-%d", p.dir, os.Getpid(), atomic.AddUint64(&p.fileNameCounter, 1))
+	log.Debug("Unified Sorter: trying to create file backEnd")
+	ret, err := newFileSorterBackEnd(fname, &msgPackGenSerde{})
+	if err != nil {
+		return nil, errors.AddStack(err)
+	}
+
+	atomic.AddInt64(&p.memoryUseEstimate, heapSizeLimit)
+	return ret, nil
+}
+
+func (p *backEndPool) dealloc(backEnd sorterBackEnd) error {
+	err := backEnd.reset()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	switch b := backEnd.(type) {
+	case *memorySorterBackEnd:
+		atomic.AddInt64(&p.memoryUseEstimate, -heapSizeLimit)
+		// Let GC do its job
+		return nil
+	case *fileSorterBackEnd:
+		for i := range p.cache {
+			ptr := &p.cache[i]
+			if atomic.CompareAndSwapPointer(ptr, nil, unsafe.Pointer(b)) {
+				return nil
+			}
+		}
+		// Cache is full.
+		err := b.free()
+		if err != nil {
+			return errors.AddStack(err)
+		}
+		return nil
+	default:
+		log.Fatal("backEndPool: unexpected backEnd type to be deallocated", zap.Reflect("type", reflect.TypeOf(backEnd)))
+	}
+	return nil
+}
+
+type flushTask struct {
+	heapSorterID  int
+	backend       sorterBackEnd
+	maxResolvedTs uint64
+	finished      chan error
+	dealloc       func() error
+	lastTs        uint64 // for debugging TODO remove
+}
+
+type heapSorter struct {
+	id          int
+	inputCh     chan *model.PolymorphicEvent
+	outputCh    chan *flushTask
+	heap        sortHeap
+	backEndPool *backEndPool
+}
+
+func newHeapSorter(id int, pool *backEndPool, out chan *flushTask) *heapSorter {
+	return &heapSorter{
+		id:          id,
+		inputCh:     make(chan *model.PolymorphicEvent, 1024*1024),
+		outputCh:    out,
+		heap:        make(sortHeap, 0, 65536),
+		backEndPool: pool,
+	}
+}
+
+// flush should only be called within the main loop in run().
+func (h *heapSorter) flush(ctx context.Context, maxResolvedTs uint64) error {
+	isEmptyFlush := h.heap.Len() == 0
+	var backEnd sorterBackEnd = nil
```

Comment:
```suggestion
	var backEnd sorterBackEnd
```

It's nil by default.